numpy/core/src/multiarray/compiled_base.c

```diff
@@ -183,39 +184,72 @@ arr_bincount(PyObject *NPY_UNUSED(self), PyObject *const *args,
         }
     }
     if (weight == Py_None) {
-        ans = (PyArrayObject *)PyArray_ZEROS(1, &ans_size, NPY_INTP, 0);
+        ans = (PyArrayObject *)PyArray_ZEROS(1, &ans_size, NPY_UINTP, 0);
         if (ans == NULL) {
             goto fail;
         }
-        ians = (npy_intp *)PyArray_DATA(ans);
+        ians = (npy_uintp *)PyArray_DATA(ans);
         NPY_BEGIN_ALLOW_THREADS;
         for (i = 0; i < len; i++)
             ians[numbers[i]] += 1;
         NPY_END_ALLOW_THREADS;
         Py_DECREF(lst);
     }
     else {
-        wts = (PyArrayObject *)PyArray_ContiguousFromAny(
-                                                weight, NPY_DOUBLE, 1, 1);
+        wts = (PyArrayObject *)PyArray_FromAny(
+            weight, NULL, 1, 1, NPY_ARRAY_DEFAULT, NULL);
         if (wts == NULL) {
             goto fail;
         }
-        weights = (double *)PyArray_DATA(wts);
         if (PyArray_SIZE(wts) != len) {
             PyErr_SetString(PyExc_ValueError,
                     "The weights and list don't have the same length.");
             goto fail;
         }
-        ans = (PyArrayObject *)PyArray_ZEROS(1, &ans_size, NPY_DOUBLE, 0);
-        if (ans == NULL) {
+        if (PyArray_ISINTEGER(wts)) {
+            iweights = (npy_uintp *)PyArray_DATA(wts);
```

Comment:
This is cast is wrong.  You should stick with `intp` everywhere.

One reason I haven't looked at this much yet is that I am not sure how far we actually want to go here, since `np.add.at` is faster now and is often a work-around (it doesn't auto-resize create/resize the result mainly).

Expanding this to more than just complex requires either proper templating (moving to `c.src` format or `C++`) or figuring out how to re-use the `ufunc.at` logic (which may well be easier, since that has fast loops now; but that needs some new logic like chunking the index to resize, which might affect performance, although I doubt much).

@mattip do you have a tought about this?